File: LEETCODE/LEETCODE-CPP/323.cpp
// 323. Number of Connected Components in an Undirected Graph
// https://leetcode.com/problems/number-of-connected-components-in-an-undirected-graph/

// Method-1 dfs
class Solution
{
public:
    int countComponents(int n, vector<vector<int>> &edges)
    {
        vector<vector<int>> adj(n);

        for (auto &edge : edges)
        {
            adj[edge[0]].push_back(edge[1]);
            adj[edge[1]].push_back(edge[0]);
        }

        int count = 0;
        vector<bool> vis(n);
        for (int i = 0; i < n; i++)
        {
            if (!vis[i])
            {
                count++;
                helper(adj, vis, i);
            }
        }

        return count;
    }

    void helper(vector<vector<int>> &adj, vector<bool> &vis, int node)
    {
        if (vis[node] == true)
        {
            return;
        }

        vis[node] = true;
        for (int neighbour : adj[node])
        {
            helper(adj, vis, neighbour);
        }
    }
};

File: LEETCODE/LEETCODE-JAVA/323.java
// 323. Number of Connected Components in an Undirected Graph
// https://leetcode.com/problems/number-of-connected-components-in-an-undirected-graph/

// Method-1 dfs
class Solution {
    public int countComponents(int n, int[][] edges) {
        List<List<Integer>> adj=new ArrayList<>();

        for(int i=0;i<n;i++){
            adj.add(new ArrayList<>());
        }

        for(int[] edge: edges){
            adj.get(edge[0]).add(edge[1]);
            adj.get(edge[1]).add(edge[0]);
        }

        boolean[] vis=new boolean[n];
        int count=0;
        for(int i=0;i<n;i++){
            if(!vis[i]){
                count++;
                helper(adj,i,vis);
            }
        }

        return count;
    }

    private void helper(List<List<Integer>> adj, int i, boolean[] vis){
        vis[i]=true;

        for(int neighbour:adj.get(i)){
            if(!vis[neighbour]){
                helper(adj,neighbour,vis);
            }
        }
    }
}